### DISCUSSION TITLE
Jolt’s preference on choosing Static body shapes

### DISCUSSION THREAD

**Discussion by PhoenixIllusion** [jrouwe/JoltPhysics.js#120] [Q&A] [answered]
On the CPP GitHub discussions, regarding static shapes and the discussion of Horizon, that it has been recommended that lots of Meshes on a height field terrain should be fine. If we are designing our scene/level with Jolt in mind, is there a preferred shape configuration we should use on Statics that is more efficient or better than this? Ex: aim for primitives, static compound shapes, us X if a user can walk inside it like a building or tunnel

Per https://github.com/jrouwe/JoltPhysics/discussions/726 - house with convex shapes, you said that static bodies should be fine with simplified meshes 

https://github.com/jrouwe/JoltPhysics/discussions/411 - limits in mesh sizes

Hypothetical Situation: a hotel building with multiple floors and multiple rooms per floor a user can walk around inside and up/down stairs to other floors 

Cases:
1. A giant single mesh
- this would result in a single body and giant AABB that would hit every single Broad and Narrow phase, but if the collision tree preserves the users location between scans, it possibly won't need to do a full tree scan each phase?

2. A mesh per floor.
- At least the broad phase may be able to exclude the floors above and below

3. Mesh per room
- similar to the above. Broad phase tree scans from scratch could exclude all but the rooms nearest to the user. 

4. Compose the room from static boxes
- this will mean a large amount of bodies into the system. This could have a big broad phase prune of hits prior to the narrow phase I believe?
- some physics systems will do planar or other geometric collision instead of triangles for boxes and spheres, which can sometimes be better than triangular detection

5. Combine a room made of boxes and other primitive into a single static compound per room 
- this would allow some reuse of similar shaped rooms, and reduce the body count, but looks like it does moving shapes into a single static compound in a body increases the AABB, and adds all the body's combined primitive collision checks to the narrow phase?

I've thought about at least splitting each floor with a single box floor/ceiling, and super low poly mesh for room walls and keeping it just to that.

I haven't run into any kind of speed issues yet with lots of static bodies, I'm just curious about what is the "ideal" for the engine to handle.

**Comment by jrouwe** [comment, marked as answer]
> but if the collision tree preserves the users location between scans, it possibly won't need to do a full tree scan each phase?

That doesn't happen. There is no such cache.

> I'm just curious about what is the "ideal" for the engine to handle.

I think this question is very hard to answer in general. Intuitively I would say 2 or 3 are the best solutions, but you really have to profile this.
